### ISSUE TITLE
Standardize srcElement, returnValue, and window.event

### ISSUE THREAD

**Issue by annevk** [whatwg/dom#334]
Previously: https://www.w3.org/Bugs/Public/show_bug.cgi?id=20713 & https://www.w3.org/Bugs/Public/show_bug.cgi?id=25002.

`event.srcElement`: this is simply an alias to `event.target`.

`event.returnValue`: also relatively simple, although `BeforeUnloadEvent` will have to override it or maybe remove its variant:

```
bool legacyReturnValue() const { return !defaultPrevented(); }
void setLegacyReturnValue(bool returnValue) { setDefaultPrevented(!returnValue); }
```

`window.event`: I'm not entirely sure how this works. E.g.,

``` js
var d = document.implementation.createHTMLDocument("x")
var x = d.createElement("x")
x.onclick = e => w(window.event)
x.click()
```

in Live DOM viewer still ends up logging an event. I guess we're just associating the event with the [current global object](https://html.spec.whatwg.org/#current-global-object) for the duration of dispatch? IDL can be `[Replaceable] readonly attribute any event;` as it needs to return undefined by default.

**Constants**: @foolip mentions there might be some more extensions that only Gecko does not implement.

I'm raising this in part because @miketaylr told me Gecko was leaning towards implementing these due to some compatibility issues and in part because I wonder if `window.event` is problematic for shadow trees in that it leaks information similar to `document.currentScript` (which we patched). @smaug----? If `window.event` is problematic we should probably attempt to fix the existing implementations to make them leak the event less.

**Comment by miketaylr** [comment]
> I'm raising this in part because @miketaylr told me Gecko was leaning towards implementing these due to some compatibility issues

Some of the broken sites are listed as dupes/blocked by https://bugzilla.mozilla.org/show_bug.cgi?id=218415 (WONTFIX'd 13 years ago...). There are more -- probably worth me digging through my bugmail and linking the rest.

**Comment by miketaylr** [comment]
`srcElement` bugs: https://bugzilla.mozilla.org/show_bug.cgi?id=453968

**Comment by cvrebert** [comment]
A couple findings from running some `returnValue` tests:

* In Chrome and Safari Tech Preview, which implement both passive event listeners and `returnValue`, setting `{passive: true}` still allows the event to be canceled using `e.returnValue = false`. (Though I only tested a custom event, not a scroll event.)
  * This seems like an oversight which ought to be addressed as part of the spec.
* In Chrome, Safari, and Edge, setting `e.returnValue = false` appears to set `e.defaultPrevented = true` even when `e.cancelable = false`.
  * Which is to say that [this assertion fails](https://github.com/w3c/web-platform-tests/blob/778f2b4cd15ee747b6b38cef856f2f5c53cd3cd4/dom/events/Event-defaultPrevented.html#L22) when the test is switched from `preventDefault` to `returnValue`.
  * This would be contrary to the "nice" route of speccing `returnValue` as a simple alias of `preventDefault`, and is a wacky behavior in general. Hopefully no code actually relies on this...

**Comment by annevk** [comment]
@RByers apparently `returnValue` lacks the "in passive listener flag" conditional. That seems like a bug, no?

**Comment by cvrebert** [comment]
Draft of `returnValue` tests: https://github.com/w3c/web-platform-tests/pull/4305

**Comment by RByers** [comment]
> @RByers apparently returnValue lacks the "in passive listener flag" conditional. That seems like a bug, no?

I could believe it. @datpuska owns this.  Dave can you take a look?

**Comment by dtapuska** [comment]
do we know how often assigning the return value from true to false occurs?

I'd prefer if the implementation was essentially pollyfill-able;

For example:
bool legacyReturnValue() const { return !defaultPrevented(); }
void setLegacyReturnValue(bool returnValue) { if (returnValue) preventDefault(); }

Which specifically disallows the true->false transition. (much like the setCancelBubble debate)

**Comment by cvrebert** [comment]
> do we know how often assigning the return value from false to true occurs?

@dtapuska This metric perhaps?:
https://www.chromestatus.com/metrics/feature/timeline/popularity/352

For `cancelBubble`, someone crafted [a metric for whether `cancelBubble` actually had an effect on the event's dispatch.](https://www.chromestatus.com/metrics/feature/timeline/popularity/1348) There doesn't seem to be an analogous one for `returnValue`, though I'm only looking at the metric names here.

**Comment by cvrebert** [comment]
> apparently returnValue lacks the "in passive listener flag" conditional.

Separate live testcase for convenience:
http://jsbin.com/homanetibo/1/edit?html,output

**Comment by dtapuska** [comment]
@cvrebert That metric tracks also true -> true calls.

**Comment by cvrebert** [comment]
☹️

**Comment by foolip** [comment]
Regarding the constants, we have these in Blink's Event.idl:
```WebIDL
    const unsigned short MOUSEDOWN = 1;
    const unsigned short MOUSEUP   = 2;
    const unsigned short MOUSEOVER = 4;
    const unsigned short MOUSEOUT  = 8;
    const unsigned short MOUSEMOVE = 16;
    const unsigned short MOUSEDRAG = 32;
    const unsigned short CLICK     = 64;
    const unsigned short DBLCLICK  = 128;
    const unsigned short KEYDOWN   = 256;
    const unsigned short KEYUP     = 512;
    const unsigned short KEYPRESS  = 1024;
    const unsigned short DRAGDROP  = 2048;
    const unsigned short FOCUS     = 4096;
    const unsigned short BLUR      = 8192;
    const unsigned short SELECT    = 16384;
    const unsigned short CHANGE    = 32768;
```

But they make no sense, they're not used for anything. Would be happy to see them in an historical.html test to provide some small incentive for them to be removed.

**Comment by foolip** [comment]
Thanks for the test! I've filed https://crbug.com/674944

**Comment by RByers** [comment]
[returnValue bug in chromium](https://bugs.chromium.org/p/chromium/issues/detail?id=277851)

**Comment by miketaylr** [comment]
I'm gonna take a stab at speccing this out. I'll start with some wpt tests for `srcElement` and `window.event`.

**Comment by miketaylr** [comment]
https://www.w3.org/Bugs/Public/show_bug.cgi?id=25002#c5 mentions that Edge re-added `event.returnValue`, but I'm not aware of any bustage in Gecko around this, while`window.event` and `window.srcElement` break a good number of sites for us.

@jacobrossi anything you can share about why it was re-added to Edge?

**Comment by miketaylr** [comment]
That said, `returnValue` isn't a huge feature and has the potential to fix unknown bustage in Gecko, so... nevermind. ^__^

**Comment by domenic** [comment]
I wonder if we should bother trying to get rid of BeforeUnloadEvent given that it's only purpose on top of Event is to add returnValue...

**Comment by miketaylr** [comment]
> I wonder if we should bother trying to get rid of BeforeUnloadEvent given that it's only purpose on top of Event is to add returnValue...

So if setting `BeforeUnloadEvent.returnValue` is [pointless in Safari, Firefox, and Chrome...](https://www.chromestatus.com/feature/5349061406228480) I don't see much of a reason for it to exist as its own interface at this point. (Note that Edge still shows script-set `returnValue` in their confirmation dialog).

Step 8 of [prompt to unload a document,](https://html.spec.whatwg.org/multipage/browsers.html#prompt-to-unload-a-document) says that `returnValue` set to the empty string is effecitvely `e.preventDefault()`, although quick testing shows that only Firefox treats `returnValue = ""` as `preventDefault()`: https://miketaylr.com/bzla/beforeunload-2.html

So...we could nuke `BeforeUnloadEvent` and treat `Event.returnValue` as follows:  

If you set `returnValue` to `false` (or the empty string for an event of type `beforeunload`), prevent the default action.
Otherwise, just set the value.

@cvrebert does that match your testing?

**Comment by cvrebert** [comment]
@miketaylr I haven't tested `beforeunload`, nor non-boolean values.

**Comment by miketaylr** [comment]
OK, thanks. Something for further testing.

**Comment by miketaylr** [comment]
> If you set returnValue to false (or the empty string for an event of type beforeunload), prevent the default action.

I did some digging on `returnValue = ""`, which is [mentioned in HTML](https://html.spec.whatwg.org/multipage/browsers.html#prompt-to-unload-a-document)

> ...and the returnValue attribute of the event object is not the empty string...

Using this test: https://miketaylr.com/bzla/beforeunload-2.html, Firefox and Presto are the only engines that don't prompt the user if you manually set `returnValue` to the empty string. @patrickkettner [tested IE6 - 8](https://twitter.com/patrickkettner/status/834074987694743553) and they behave like Gecko and Presto.

OTOH, Chrome, Safari, (new) Opera, Edge and IE9+ do show the prompt. 

Since IE9+ and Edge didn't require this for webcompat, *and* WebKit/Blink never had it, that's a pretty good signal we could remove that special handling from HTML and not bother with it here.

(I wonder if IE maintained that behavior for attachEvent, or something. But that doesn't matter).

**Comment by patrickkettner** [comment]
> I wonder if IE maintained that behavior for attachEvent, or something. But that doesn't matter

just checked. nope.

**Comment by domenic** [comment]
`returnValue` and beforeunload is kind of messy. See some recent fixes for it in https://github.com/whatwg/html/pull/2353. /cc @cdumez, who is landing patches to align WebKit with the spec/Gecko in that regard.

My understanding is that the HTML spec is currently based on a world where returnValue does not exist on Event, and does exist on BeforeUnloadEvent. Gecko, and WebKit as of @cdumez's recent changes, implement that part of the HTML spec percectly; Chrome does so imperfectly, see https://bugs.chromium.org/p/chromium/issues/detail?id=692203. (I'm not sure about Edge because I cannot run tests on beforeunload there due to their longstanding issues where the outer frame has issues reaching into other iframes; see e.g. their inability to pass https://github.com/w3c/web-platform-tests/pull/4800 tests.)

I think it's possible we could try to aim for a world where all the returnValue weirdness lives in the base Event class, and BeforeUnloadEvent perhaps even disappears entirely. But we should outline concretely what that looks like, and get as many vendors as possible on board with such a plan, before causing more churn.

**Comment by domenic** [comment]
Hmm. I guess I forgot that there are two separate pieces of weirdness: manually setting the `returnValue`, and handling the return value of an `onbeforeunload` event handler. @miketaylr is mostly talking about the former, whereas my post and the recent changes were mostly about the latter.

Now that I remember that, it seems like moving `returnValue` into the base `Event` class is not as churn-ful as I thought it would be, so I am less concerned.

But still we should try to outline a concrete proposal (i.e.: what spec lines change) and see what browsers currently depart from it and whether they'd be willing to change.

**Comment by cdumez** [comment]
Slight weirdness in WebKit, returnValue property is both on BeforeUnloadEvent and the base Event (likely for backward compatibility reasons from when we did not have BeforeUnloadEvent).

**Comment by miketaylr** [comment]
Yeah, let's try to avoid any churn (I'm try to be careful here...).

My thinking here: 

`Event.returnValue`:

* If you set `returnValue` to `false`, prevent the default action.
* Otherwise, just set the value.

(I haven't tested existing handling of `0` or `null` yet, on my TODO list.)

> Gecko, and WebKit as of @cdumez's recent changes, implement that part of the HTML spec perfectly;

Maybe I misunderstand (odds are good...): https://miketaylr.com/bzla/beforeunload.html in WebKit shows the beforeunload prompt when `returnValue` is set to the empty string, which is not what HTML says. So currently the only maintained browser that has that behavior is Gecko. I suspect it's not needed.

`BeforeUnload.returnValue` at that point just behaves like `Event`, so I think we can get rid of it.

I'd prefer to not touch any other aspect of `beforeunload` event handling at all, if possible.

**Comment by cdumez** [comment]
@miketaylr In your example, I believe WebKit nightly shows a confirmation dialog only because you return e in your event handler, which gets converted to a non-empty string, NOT because the returnValue being set to an empty string.

WebKit ToT matches the spec and Gecko in this regard.

**Comment by domenic** [comment]
In Chrome at least Event's returnValue is a Web IDL boolean, so setting it to 0 and null should be the same as setting it to false.

It sounds like your proposing one of the following deltas from the current specs:

1. Add returnValue as a boolean to Event, so that setting it to false (or something that converts to false) cancels the event.
2. (1), but also:
   - Remove returnValue as a DOMString from BeforeUnloadEvent, making it an empty interface
   - Remove the setting of `e.returnValue` to the return value of `onbeforeunload` event handlers in https://html.spec.whatwg.org/#the-event-handler-processing-algorithm step 4
   - Remove the checking of `e.returnValue` in https://html.spec.whatwg.org/#unloading-documents:dom-beforeunloadevent-returnvalue (thus relying entirely on the checking of cancelation)
3. (2), but also remove the BeforeUnloadEvent interface entirely

**Comment by miketaylr** [comment]
> @miketaylr In your example, I believe WebKit nightly shows a confirmation dialog only because you return e in your event handler, which gets converted to a non-empty string, NOT because the returnValue being set to an empty string.

Ah, could be. Will tweak and re-test when I get some more time this afternoon.

**Comment by miketaylr** [comment]
> WebKit ToT matches the spec and Gecko in this regard.

Yeah, verified w/ https://miketaylr.com/bzla/beforeunload-3.html (where `e` isn't returned`). But it doesn't match in https://miketaylr.com/bzla/beforeunload-2.html (where `e` is returned).

@cdumez was there any motivation to move WebKit to this behavior (other than spec compliance)? If there are sites that depend on it, that would be useful information, especially given the split in behavior between Gecko/WebKit (Trunk)/Presto and Blink/Edge/IE.

**Comment by cdumez** [comment]
@miketaylr We made the change to align with the specification and Firefox, and because @domenic asked us via https://bugs.webkit.org/show_bug.cgi?id=168382. I am not aware of any site that depends on this behavior.

**Comment by domenic** [comment]
@miketaylr I'm still confused which of the three options in https://github.com/whatwg/dom/issues/334#issuecomment-281404452 you are hoping to move us toward?

**Comment by miketaylr** [comment]
> @miketaylr I'm still confused which of the three options in #334 (comment) you are hoping to move us toward?

I was originally thinking 3, but that seems less desirable with WebKit having just matched Gecko's behavior.

**Comment by cdumez** [comment]
Well, WebKit has not shipped this yet and there is still time for us to make changes if needed. Do not let my recent changes discourage you from doing the right thing spec-wise. My personal opinion is that we should choose the path that requires the less changes in all browsers to achieve interoperability and spec that (Note that this is not necessarily the cleanest solution). I have bad experience with the spec trying to "do the nicest / clean thing" but browsers never aligning.

**Comment by domenic** [comment]
My perspective:

- (2) is less likely to be web-compatible than (1), but the visible changes are so small compared to today's intersection semantics, I very much doubt it would cause problems. (Basically, people would have to be reading back the value of `e.returnValue` and expecting a string, not just a truthy value.)
- (3) seems a bit more dangerous.
- (2) or (3) require many more changes in browsers to achieve than (1).

So on balance probably (1) is going to be the way to go, even if it's significantly less nice.

**Comment by miketaylr** [comment]
> So on balance probably (1) is going to be the way to go, even if it's significantly less nice.

Less nice, but less risky. I'm game.

**Comment by cdumez** [comment]
It sounds like I have nothing to do in WebKit if we go with (1) so I like it 😃

**Comment by annevk** [comment]
This was raised earlier, but do we also make it so that `returnValue` can only be set once? I.e., no setting it false and then true? I believe that would require changes to implementations, but would also make it so that it's not a more powerful API than what we have thus far.

**Comment by foolip** [comment]
That sounds similar to what was done for [`cancelBubble`](https://dom.spec.whatwg.org/#dom-event-cancelbubble). https://www.chromestatus.com/metrics/feature/timeline/popularity/352 was linked [above](https://github.com/whatwg/dom/issues/334#issuecomment-266086465), but it doesn't quite answer how risky this would be. However, it seems likely that a majority of that usage is going from false to true, as that much no-oping would be odd.

I wonder though if such code is actually trying to stop canceling the event, or setting a string or something.

**Comment by miketaylr** [comment]
*dusts off issue*

> I wonder if window.event is problematic for shadow trees in that it leaks information similar to document.currentScript (which we patched). @smaug----? If window.event is problematic we should probably attempt to fix the existing implementations to make them leak the event less.

@annevk @smaug----  it doesn't look like it, with some basic testing here: https://miketaylr.com/bzla/global-event-shadow.html (open devtools console and click on the "click me" element).

Anything else obvious I should be looking at?

**Comment by smaug----** [comment]
What happens if script in shadow DOM calls some non-shadow method which accessed window.event? Does window.event.target point to a node in shadow DOM?

**Comment by miketaylr** [comment]
> What happens if script in shadow DOM calls some non-shadow method which accessed window.event? Does window.event.target point to a node in shadow DOM?

in https://miketaylr.com/bzla/global-event-shadow-script.html a click listener is added inside the shadow, but points to a non-shadow listener callback `nonShadowGetTarget`. Both `e.target` and `event.target` point to the same non-shadow nodes.

**Comment by annevk** [comment]
@miketaylr when explicitly leaking the event object outside the shadow tree it's reasonable for things to leak. The problem is that if you just call a global function from the event listener even though it doesn't have access to the event normally, it will have access to `window.event`, which is a problem of sorts. I suggest we basically don't set `window.event` when the event target is inside a shadow tree. That should be compatible enough since shadow trees are a new feature.

**Comment by miketaylr** [comment]
> when explicitly leaking the event object outside the shadow tree it's reasonable for things to leak. The problem is that if you just call a global function from the event listener even though it doesn't have access to the event normally, it will have access to window.event, which is a problem of sorts.

@annevk so far my tests show that nothing leaking from the shadow tree (which maybe isn't obvious given the way I wrote them), in both instances `event.target` and `window.event` point to the light dom wrapper node (or whatever the right term is).

**Comment by annevk** [comment]
@miketaylr I think that's only the case because you didn't add a listener inside the shadow tree.

**Comment by miketaylr** [comment]
> @miketaylr I think that's only the case because you didn't add a listener inside the shadow tree.

Yeah, you're right. Updated https://miketaylr.com/bzla/global-event-shadow-script.html and the results are a bit confusing (possibly due to devtools bugs?...). 

In Chrome, nothing gets logged except the `console.group` names, but in Safari `event.target.tagName` and `e.target.tagName` are `SPAN`, which means it leaked.

So we should probably set `window.event` to undefined when traversing across shadow boundaries (or something like that).

**Comment by miketaylr** [comment]
> So we should probably set window.event to undefined when traversing across shadow boundaries (or something like that).

This seems to do that: 

https://github.com/whatwg/dom/commit/9ada239e14fe7964577276ef9d8972069e665247

**Comment by miketaylr** [comment]
Back to working on this this week.

@hayatoito could you point out where in Blink `window.event` is set? Presumably somewhere during event dispatch, but I'm unfamiliar with Blink code. Thanks.

**Comment by hayatoito** [comment]
@yuki3 and @peria,

Could you help me to understand where Blink set `window.event`?
I tried to find it, but it looks v8 (or binding layers) set (or get) it in a unfamiliar way to me. :(

- https://cs.chromium.org/chromium/src/third_party/WebKit/Source/core/frame/Window.idl?q=Window.idl&sq=package:chromium&l=179
- https://cs.chromium.org/chromium/src/third_party/WebKit/Source/bindings/core/v8/custom/V8WindowCustom.cpp?l=103

**Comment by peria** [comment]
`window.event` is set on a `DOMWindow` object in [`V8Window::eventAttributeSetterCustom()`](https://cs.chromium.org/chromium/src/third_party/WebKit/Source/bindings/core/v8/custom/V8WindowCustom.cpp?l=119), and it is called from a V8 routine through [`V8Window::eventAttributeSetterCallback`](https://cs.chromium.org/chromium/src/out/Debug/gen/blink/bindings/core/v8/V8Window.cpp?l=6640), but I'm not sure when it happens. :(

**Comment by yuki3** [comment]
In Blink, [`V8AbstractEventListener::InvokeEventHandler`](https://cs.chromium.org/chromium/src/third_party/WebKit/Source/bindings/core/v8/V8AbstractEventListener.cpp?q=V8AbstractEventListener::InvokeEventHandler&sq=package:chromium&l=120) does the trick.  Roughly speaking, every time that Blink dispatches an `Event`, Blink puts each `Event` on stack (note that it's possible that event handling is nested), and the latest `Event` is associated with global object's `event` attribute (internally Blink is using a V8's private property for this purpose).  This is true not only for Window but also Worker/Worklet.

In [`V8AbstractEventListener::InvokeEventHandler`](https://cs.chromium.org/chromium/src/third_party/WebKit/Source/bindings/core/v8/V8AbstractEventListener.cpp?q=V8AbstractEventListener::InvokeEventHandler&sq=package:chromium&l=120), you see `saved_event` and `V8PrivateProperty::GetGlobalEvent`.

**Comment by hayatoito** [comment]
@peria, @yuki3, thank you!

I chatted with @yuki3 offline. From a spec perspective, here is the summary, as far as I understand:
- Blink sets `window.event` to an `Event` just before calling a registered event listener.
- Blink sets `window.event` to `undefined` after an event listener finishes.
- Since users can dispatch an event synchronously in an event listener, Blink uses *stack* so that this works in *reentrant* way.

**Comment by smaug----** [comment]
But it doesn't deal with Shadow DOM in any reasonable/special way?

**Comment by hayatoito** [comment]
Yup, Blink doesn't have any special treatment for Shadow DOM, regarding `window.event`. `window.event` returns an Event without any checking for Shadow DOM encapsulation. That is the current status.

I remember we had a similar discussion for `document.currentScript`.
Would it be better to make window.Event return `undefined` if an event's target is in a shadow tree?

I think that's okay. Blink would have an incompatibility issue by this change, however, I think that is unlikely to happen, hopefully.

**Comment by annevk** [comment]
@hayatoito yeah, I think that's exactly what we want to do here.

**Comment by hayatoito** [comment]
I see. Let me try to update Blink's behavior after we standardize `window.event` as such.

**Comment by miketaylr** [comment]
Thank you @hayatoito @yuki3 @peria, this is very helpful.

**Comment by RByers** [comment]
FWIW I agree that we should just try to break this in Blink - the compat risk is likely low.  @hayatoito, feel free to propose a a change / "intent to remove" whenever you like based on this discussion, I don't think we need to wait for the spec to be ready if we all agree that's the plan (especially if you can help Mike out with his web-platform-tests for this).

**Comment by hayatoito** [comment]
Yup, I agree that we don't need to wait for it. Let me send "Intent to remove window.event for Shadow DOM" or something to blink-dev.

**Comment by miketaylr** [comment]
> do we know how often assigning the return value from false to true occurs?
> 
> I'd prefer if the implementation was essentially pollyfill-able;
> 
> For example:
> bool legacyReturnValue() const { return !defaultPrevented(); }
> void setLegacyReturnValue(bool returnValue) { if (!returnValue) preventDefault(); }
> 
> Which specifically disallows the false->true transition. (much like the setCancelBubble debate)

The current spec-text in my PR doesn't say anything about this, but @cvrebert wrote a wpt test to ensure that setting `returnValue` to true after `preventDefault()` was called should no-op. 

Safari, Chrome and Edge will all unset the canceled flag, however: 
https://miketaylr.com/bzla/canceled-returnValue-false-true.html

@dtapuska do we want to spec reality here, or stick with your original suggestion (which means I should add some language to ensure `returnValue=true` does nothing if the canceled flag was set).

**Comment by annevk** [comment]
Some issues with `window.event`:

* Chrome has (non-standard) [NotEnumerable] in IDL meaning `Object.getOwnPropertyDescriptor()` returns the wrong thing there. Is that something that can be removed by Chrome or do we need it in the standard?
* It's a little bit unclear where the global is gotten from. I thought it was related to the target/currentTarget, but https://cs.chromium.org/chromium/src/third_party/WebKit/Source/bindings/core/v8/V8AbstractEventListener.cpp?q=V8AbstractEventListener::InvokeEventHandler&sq=package:chromium&l=135 suggests it's global state. Does that correspond to the "current global object" concept or is it something else? From the test cases I assumed it was somehow related to currentTarget, since if you move nodes in the path between documents you can get the global whose `.event` is set to change as well. But maybe that's just how callbacks work and I'm a little confused.

**Comment by annevk** [comment]
Another thing I can't quite explain in Chrome is that if you have 2 nodes and the parent node has two capture listeners and the child node has a non-capture listener. Then in the first parent node listener you move the node to another document. In this listener the global is still the original. In the second listener is still the original, but in the child node it's the global of the new document. I don't get why it hasn't changed in the second parent node listener. ("window.event and moving an element post-dispatch" in https://github.com/w3c/web-platform-tests/pull/10329 tests this scenario. The asserts for that test are written to match Chrome.)

**Comment by yuki3** [comment]
For [NotEnumerable] for window.event, I have no evidence, but I guess that `window.event` was not standardized, so Blink didn't want to list up the property in case of enumeration, i.e. Blink wanted to make it a kind of a hidden property because the spec expected Blink to not have window.event.  I think that we can remove [NotEnumerable] as window.event is now spec'ed.

For which global object Blink is using, I'm not 100% sure, but it looks like Blink is using the global object associated with the EventTarget.  I tried to track down where the global object comes from, and reached here:
https://cs.chromium.org/chromium/src/third_party/WebKit/Source/core/dom/events/EventTarget.cpp?rcl=8f0f9ceb2c3f5acf36ef55d620d3f0e6ea7d8d01&l=748
My code reading could be wrong, but I guess that Blink uses the EventTarget => its Document (or Worker) => its Window (or Worker).

For node movement during event dispatch, I know nothing, but I think that it's possible that the behavior was implemented without deep thoughts.  Maybe Blink should fix it, too.

**Comment by dmethvin** [comment]
It seems like [V8AbstractEventListener.cpp](https://cs.chromium.org/chromium/src/third_party/WebKit/Source/bindings/core/v8/V8AbstractEventListener.cpp?q=V8AbstractEventListener::InvokeEventHandler&sq=package:chromium&l=120) saves `window.event` before invoking the listener and then always restores it to the old value on exit. Any change to `window.event` that resulted from the event handler being called would be clobbered.

**Comment by yuki3** [comment]
`window.event` is currently implemented as [Replaceable] in Blink, so, if author script overrides it, then we always return the value set by author script.

**Comment by smaug----** [comment]
Not always. While events are being handled, window.event returns the Event object.
`data:text/html,<script>window.event = "value before event dispatch"; window.addEventListener("foo", function(e) { console.log("window.event in event listener: " + window.event); });  window.dispatchEvent(new Event("foo")); console.log("window.event after event dispatch: " + window.event);</script> `

**Comment by yuki3** [comment]
@smaug---- , I copy&pasted your code and it showed:
```
window.event in event listener: value before event dispatch
window.event after event dispatch: value before event dispatch
```
So, [Replaceable] seems working just fine to me.

Even if this were not working as I described, it should be an issue of [Replaceable] rather than of `window.event`.

**Comment by smaug----** [comment]
My point was just that "if author script overrides it, then we always return the value set by author script." wasn't quite right. When the listener is called, the value set by author isn't returned.

**Comment by annevk** [comment]
@smaug---- it is?

**Comment by smaug----** [comment]
Apparently I'm lost now, and made some mistake in testing.

**Comment by annevk** [comment]
@yuki3 the main thing I don't understand is the result of https://github.com/whatwg/dom/issues/334#issuecomment-379247567 in Chrome. If you move an element around it doesn't affect subsequent listeners for that element somehow, but it does affect any of its children. However, Chrome appears to lookup the global for each listener and doesn't cache it across listeners, so what's going on?

Having said that, could we maybe get agreement on a different setup given that it mostly seems arbitrary? Where we lookup the global once and then use it throughout? This would be more consistent with other bits that we also try to lookup and cache in advance.

(Edge seems to update multiple windows once you move a node, but not before, which seems undesirable.)

cc @dstorey

**Comment by annevk** [comment]
Well, I don't really want to support it, but every browser other than Firefox having it makes it somewhat essential unfortunately. I guess that would be a reasonable alternative model, except that I think I learned from @domenic that functions in ECMAScript are not supposed to have an associated global (even though this is the case in some implementations).

**Comment by domenic** [comment]
Nah, functions always have an associated global, even in the ES spec. It's non-function objects where we have the implementation/spec mismatch.

(I don't understand the details of window.event well enough to comment on the larger question.)

**Comment by annevk** [comment]
An event listener can be a non-function object, but it seems that per https://heycam.github.io/webidl/#call-a-user-objects-operation and https://heycam.github.io/webidl/#dfn-associated-realm we already "monkey patch" ECMAScript to add this concept for all objects...

I guess if we're sure that's not going away and everyone is okay with using that we could, though no implementer does so currently (there's a test for that in the aforementioned PR as well).

**Comment by yuki3** [comment]
I'm fine and I think it's reasonable to use the listener's associated global.

**Comment by annevk** [comment]
I've updated the PR with that suggestion.

**Comment by dstorey** [comment]
Edge has the exact same behavior as mentioned by @dmethvin above
